### PR TITLE
Tests for collision methods and fix bugs in return values

### LINKED ISSUE #9: Corner Collisions triggering touching.left
Created a simple map using an array to generate blocks.  When running on a long string of blocks, touching.right worked fine but touching.left would trigger from corner collisions.  Thus making collide() trigger and stop movement. 

[example with touching.left being logged to console](http://matthesby.github.io/p5playplatformer/)

### PR DESCRIPTION
**Pull request by islemaster**
The four collision methods in p5.play (`overlap`, `collide`, `displace` and `bounce`) had some common problems where their return values did not behave as documented when groups are involved.  I've fixed those return value problems and written fairly comprehensive test coverage of how these methods report collision when it occurs.

My tests do _not_ cover the behavior of different collider types, whether overlap is calculated correctly, or the changes to sprite position/velocity that can occur as a result of collisions (for example, I have not fixed #9 yet).  I'm only concerned here with boolean return values, and whether callbacks are called the appropriate number of times with the appropriate arguments.

# The bugs
Sprite-sprite checks seem to work as expected.

* `sprite.collide(group)`, `sprite.displace(group)` and `sprite.bounce(group)` had a problem where they would return the value of the _last_ check they did against the group, where the expected behavior is to return true if _any_ collision occurred.  `sprite.overlap(group)` did not have this problem, because it's a separate case inside the `AABBops` method.
* `group.overlap(sprite)` (and each variant) was documented to return a boolean for whether an overlap occurred, but its implementation always returned `undefined`.
* `group.overlap(group)` exhibited both of the above problems.

# The tests
The test suite is long and has quite a bit of copy-paste to cover the four similar collision methods in all four caller-callee configurations: sprite(sprite), sprite(group), group(sprite) and group(group).  I went ahead and kept much of the duplication because there _are_ some subtle differences between tests for the different methods - for example, when all sprites start in the same position, sprite.collide(group) will result in only one callback, since the first collision displaces the caller, but sprite.displace(group) will result in many callbacks as the caller displaces each group member in turn.

When reading the tests, the general setup is this:

* There are four sprites, A through D.  They all start in different positions.
* There are two groups, AB and CD, named for the sprites they contain.
* If sprites should overlap for a test they will always be explicitly positioned prior to the test by calling `moveAToB()`.

Here's the complete successful test run (from collisions.js, only):
```
  collisions
    sprite.overlap(sprite)
      ✓ false if sprites do not overlap 
      ✓ true if sprites overlap 
      ✓ calls callback once if sprites overlap 
      ✓ does not call callback if sprites do not overlap 
      passes collider and collidee to callback
        ✓ A-B 
        ✓ B-A 
    sprite.overlap(group)
      ✓ false if sprite does not overlap any sprites in group 
      ✓ does not check against sprites not in target group 
      ✓ does not care if sprites in target group overlap each other 
      ✓ true if sprite overlaps first sprite in group 
      ✓ true if sprite overlaps last sprite in group 
      ✓ true if sprite overlaps every sprite in group 
      ✓ does not overlap self when checking own group 
      ✓ can overlap with other sprites in own group 
      ✓ does not call callback when not overlapping any sprites 
      ✓ calls callback once when overlapping one sprite 
      ✓ calls callback twice when overlapping two sprites 
      passes collider and collidee to callback
        ✓ A-C 
        ✓ A-D 
        ✓ A-C-D 
    sprite.displace(sprite)
      ✓ false if sprites do not overlap 
      ✓ true if sprites overlap 
      ✓ calls callback once if sprites overlap 
      ✓ does not call callback if sprites do not overlap 
      passes collider and collidee to callback
        ✓ A-B 
        ✓ B-A 
    sprite.displace(group)
      ✓ false if sprite does not overlap any sprites in group 
      ✓ does not check against sprites not in target group 
      ✓ does not care if sprites in target group overlap each other 
      ✓ true if sprite overlaps first sprite in group 
      ✓ true if sprite overlaps last sprite in group 
      ✓ true if sprite overlaps every sprite in group 
      ✓ does not overlap self when checking own group 
      ✓ can overlap with other sprites in own group 
      ✓ does not call callback when not overlapping any sprites 
      ✓ calls callback once when overlapping one sprite 
      ✓ calls callback twice when overlapping two sprites 
      passes collider and collidee to callback
        ✓ A-C 
        ✓ A-D 
        ✓ A-C-D 
    sprite.collide(sprite)
      ✓ false if sprites do not overlap 
      ✓ true if sprites overlap 
      ✓ calls callback once if sprites overlap 
      ✓ does not call callback if sprites do not overlap 
      passes collider and collidee to callback
        ✓ A-B 
        ✓ B-A 
    sprite.collide(group)
      ✓ false if sprite does not overlap any sprites in group 
      ✓ does not check against sprites not in target group 
      ✓ does not care if sprites in target group overlap each other 
      ✓ true if sprite overlaps first sprite in group 
      ✓ true if sprite overlaps last sprite in group 
      ✓ true if sprite overlaps every sprite in group 
      ✓ does not overlap self when checking own group 
      ✓ can overlap with other sprites in own group 
      ✓ does not call callback when not overlapping any sprites 
      ✓ calls callback once when overlapping one sprite 
      ✓ calls callback twice when overlapping two sprites 
      passes collider and collidee to callback
        ✓ A-C 
        ✓ A-D 
        ✓ A-C-D 
    sprite.bounce(sprite)
      ✓ false if sprites do not overlap 
      ✓ true if sprites overlap 
      ✓ calls callback once if sprites overlap 
      ✓ does not call callback if sprites do not overlap 
      passes collider and collidee to callback
        ✓ A-B 
        ✓ B-A 
    sprite.bounce(group)
      ✓ false if sprite does not overlap any sprites in group 
      ✓ does not check against sprites not in target group 
      ✓ does not care if sprites in target group overlap each other 
      ✓ true if sprite overlaps first sprite in group 
      ✓ true if sprite overlaps last sprite in group 
      ✓ true if sprite overlaps every sprite in group 
      ✓ does not overlap self when checking own group 
      ✓ can overlap with other sprites in own group 
      ✓ does not call callback when not overlapping any sprites 
      ✓ calls callback once when overlapping one sprite 
      ✓ calls callback twice when overlapping two sprites 
      passes collider and collidee to callback
        ✓ A-C 
        ✓ A-D 
        ✓ A-C-D 
    group.overlap(sprite)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group overlaps target sprite 
      ✓ true if all sprites in group overlap target sprite 
      ✓ does not call callback when not overlapping sprite 
      true if any sprites in group overlap target sprite
        ✓ A overlaps C 
        ✓ B overlaps C 
      passes collider and collidee to callback for each overlap
        ✓ A-C 
        ✓ B-C 
        ✓ A-B-C 
    group.overlap(group)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group AB overlaps any sprite in group CD 
      ✓ true if all sprites in group AB overlap all sprites in group CD 
      ✓ group does not overlap self if no sprites within group overlap 
      ✓ group overlaps self if two different sprites within group overlap 
      ✓ does not call callback when not overlapping any sprites 
      true if any sprites in group AB overlap any sprites in group CD
        ✓ A overlaps C 
        ✓ A overlaps D 
        ✓ B overlaps C 
        ✓ B overlaps D 
      passes collider and collidee to callback for each pair
        ✓ A-C 
        ✓ A-D 
        ✓ B-C 
        ✓ B-D 
        ✓ A-B-C 
        ✓ A-B-D 
        ✓ A-C-D 
        ✓ B-C-D 
        ✓ A-C, B-D 
        ✓ A-D, B-C 
        ✓ A-B-C-D 
    group.displace(sprite)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group overlaps target sprite 
      ✓ true if all sprites in group overlap target sprite 
      ✓ does not call callback when not overlapping sprite 
      true if any sprites in group overlap target sprite
        ✓ A overlaps C 
        ✓ B overlaps C 
      passes collider and collidee to callback for each overlap
        ✓ A-C 
        ✓ B-C 
        ✓ A-B-C 
    group.displace(group)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group AB overlaps any sprite in group CD 
      ✓ true if all sprites in group AB overlap all sprites in group CD 
      ✓ group does not overlap self if no sprites within group overlap 
      ✓ group overlaps self if two different sprites within group overlap 
      ✓ does not call callback when not overlapping any sprites 
      true if any sprites in group AB overlap any sprites in group CD
        ✓ A overlaps C 
        ✓ A overlaps D 
        ✓ B overlaps C 
        ✓ B overlaps D 
      passes collider and collidee to callback for each pair
        ✓ A-C 
        ✓ A-D 
        ✓ B-C 
        ✓ B-D 
        ✓ A-B-C 
        ✓ A-B-D 
        ✓ A-C-D 
        ✓ B-C-D 
        ✓ A-C, B-D 
        ✓ A-D, B-C 
        ✓ A-B-C-D 
    group.collide(sprite)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group overlaps target sprite 
      ✓ true if all sprites in group overlap target sprite 
      ✓ does not call callback when not overlapping sprite 
      true if any sprites in group overlap target sprite
        ✓ A overlaps C 
        ✓ B overlaps C 
      passes collider and collidee to callback for each overlap
        ✓ A-C 
        ✓ B-C 
        ✓ A-B-C 
    group.collide(group)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group AB overlaps any sprite in group CD 
      ✓ true if all sprites in group AB overlap all sprites in group CD 
      ✓ group does not overlap self if no sprites within group overlap 
      ✓ group overlaps self if two different sprites within group overlap 
      ✓ does not call callback when not overlapping any sprites 
      true if any sprites in group AB overlap any sprites in group CD
        ✓ A overlaps C 
        ✓ A overlaps D 
        ✓ B overlaps C 
        ✓ B overlaps D 
      passes collider and collidee to callback for each pair
        ✓ A-C 
        ✓ A-D 
        ✓ B-C 
        ✓ B-D 
        ✓ A-B-C 
        ✓ A-B-D 
        ✓ A-C-D 
        ✓ B-C-D 
        ✓ A-C, B-D 
        ✓ A-D, B-C 
        ✓ A-B-C-D 
    group.bounce(sprite)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group overlaps target sprite 
      ✓ true if all sprites in group overlap target sprite 
      ✓ does not call callback when not overlapping sprite 
      true if any sprites in group overlap target sprite
        ✓ A overlaps C 
        ✓ B overlaps C 
      passes collider and collidee to callback for each overlap
        ✓ A-C 
        ✓ B-C 
        ✓ A-B-C 
    group.bounce(group)
      ✓ false if no sprites overlap 
      ✓ false if no sprite in group AB overlaps any sprite in group CD 
      ✓ true if all sprites in group AB overlap all sprites in group CD 
      ✓ group does not overlap self if no sprites within group overlap 
      ✓ group overlaps self if two different sprites within group overlap 
      ✓ does not call callback when not overlapping any sprites 
      true if any sprites in group AB overlap any sprites in group CD
        ✓ A overlaps C 
        ✓ A overlaps D 
        ✓ B overlaps C 
        ✓ B overlaps D 
      passes collider and collidee to callback for each pair
        ✓ A-C 
        ✓ A-D 
        ✓ B-C 
        ✓ B-D 
        ✓ A-B-C 
        ✓ A-B-D 
        ✓ A-C-D 
        ✓ B-C-D 
        ✓ A-C, B-D 
        ✓ A-D, B-C 
        ✓ A-B-C-D 
```

### CODE REVIEW

**Comment by islemaster** [comment]
Also, totally open to suggestions on reordering or somehow deduplicating tests, as long as the tests stay easy to read.  For example, the following tests are nearly identical for every collision method.

```javascript
it('group does not overlap self if no sprites within group overlap', function () {
  expect(groupAB.collide(groupAB)).to.be.false;
});

it('group overlaps self if two different sprites within group overlap', function () {
  moveAToB(spriteA, spriteB);
  expect(groupAB.collide(groupAB)).to.be.true;
});
```

The answer might be to test the underlying method `AABBops` instead in these cases - but even it has a huge switch statement and behavior diverges between the four methods in places where you wouldn't expect it.  It's possible being overly verbose in our coverage here is a benefit for this particularly tricky case.

File: lib/p5.play.js [hunk 1]
```diff
@@ -2636,6 +2632,22 @@ function Group() {
     return obj === other;
   }
 
+  /**
+   * Collide each member of group against the target using the given collision
+   * type.  Return true if any collision occurred.
+   * Internal use
```
Comment (toolness) [inline lib/p5.play.js]: If it's for internal use, should it have the `@private` tag? I'm not super familiar with YUIDoc so I don't know myself, just wondering.

File: lib/p5.play.js [hunk 2]
```diff
@@ -2235,9 +2233,7 @@ function Sprite(pInst, _x, _y, _w, _h) {
               displacement = this.collider.collide(other.collider);
 
 
-            if(displacement.x == 0 &&  displacement.y == 0 )
-              result = false;
-            else
+            if(displacement.x != 0 || displacement.y != 0 )
```
Comment (toolness) [inline lib/p5.play.js]: Should this use `!==` to be consistent with the change you made [around line 144](https://github.com/molleindustria/p5.play/pull/44/files#diff-84d79f68ce33ce29426e52fd37061b4eR2144)?

**Comment by toolness** [comment]
This is awesome! :+1: 

I think it's fine to keep the tests as they are... Your reasoning is sound and I like that the tests themselves are very compact and easy to read. Thanks for the explanations in this PR, too.

Feel free to address my minor nits if you want, or just merge straight away :grin:

File: lib/p5.play.js (hunk 1 above)
Comment (islemaster) [inline lib/p5.play.js]: Yeah, I'll go ahead and do that.  There's a weird behavior in the documentation where private methods show up in the table of contents even if you don't have the option to show private methods checked, but I'll go ahead and doc this.

File: lib/p5.play.js (hunk 2 above)
Comment (islemaster) [inline lib/p5.play.js]: Nice catch, thanks.

**Pull request merged**